plugins/out_azure_logs_ingestion/azure_logs_ingestion.c

```diff
@@ -0,0 +1,445 @@
+        msgpack_unpacked_destroy(&result);
+        return -1;
+    }
+
+    msgpack_sbuffer_destroy(&mp_sbuf);
+    msgpack_unpacked_destroy(&result);
+
+    *out_buf = record;
+    *out_size = flb_sds_len(record);
+
+    return 0;
+}
+
+/* Gets OAuth token; (allocates sds string) */
+flb_sds_t get_az_li_token(struct flb_az_li *ctx)
+{
+    int ret = 0;
+    char* token;
+    size_t token_len;
+
+    if (pthread_mutex_lock(&ctx->token_mutex)) {
+        flb_plg_error(ctx->ins, "error locking mutex");
+        return NULL;
+    }
+    /* Retrieve access token only if expired */
+    if (flb_oauth2_token_expired(ctx->u_auth) == FLB_TRUE) {
+        flb_plg_debug(ctx->ins, "token expired. getting new token");
+        /* Clear any previous oauth2 payload content */
+        flb_oauth2_payload_clear(ctx->u_auth);
+
+        ret = flb_oauth2_payload_append(ctx->u_auth, "grant_type", 10, 
+                                        "client_credentials", 18);
+        if (ret == -1) {
+            flb_plg_error(ctx->ins, "error appending oauth2 params");
+            return NULL;
+        }
+
+        ret = flb_oauth2_payload_append(ctx->u_auth, "scope", 5, FLB_AZ_LI_AUTH_SCOPE,
+                                        sizeof(FLB_AZ_LI_AUTH_SCOPE) - 1);
+        if (ret == -1) {
+            flb_plg_error(ctx->ins, "error appending oauth2 params");
+            return NULL;
+        }
+
+        ret = flb_oauth2_payload_append(ctx->u_auth, "client_id", 9,
+                                        ctx->client_id, -1);
+        if (ret == -1) {
+            flb_plg_error(ctx->ins, "error appending oauth2 params");
+            return NULL;
+        }
+
+        ret = flb_oauth2_payload_append(ctx->u_auth, "client_secret", 13,
+                                        ctx->client_secret, -1);
+        if (ret == -1) {
+            flb_plg_error(ctx->ins, "error appending oauth2 params");
+            return NULL;
+        }
+
+        token = flb_oauth2_token_get(ctx->u_auth);
+
+        /* Copy string to prevent race conditions */
+        if (!token) {
+            flb_plg_error(ctx->ins, "error retrieving oauth2 access token");
+            return NULL;
+        }
+        else {
+            token_len = flb_sds_len(ctx->u_auth->token_type) + 2 +
+                                    flb_sds_len(ctx->u_auth->access_token);
+            /* If token string isn't allocated */
+            if (!ctx->token) {
+                ctx->token = flb_sds_create_size(token_len);
+                if (!ctx->token) {
+                    flb_plg_error(ctx->ins, "error creating token buffer");
+                    return NULL;
+                }
+            }
+            /* If our token_length is more than what we already allocated */
+            else if (token_len > flb_sds_len(ctx->token)) {
+                flb_plg_debug(ctx->ins, "new token len > previous token len");
+                ctx->token = flb_sds_increase(ctx->token, token_len - 
```

Comment:
You need to use a temporary variable here to save the result of `flb_sds_increase`, otherwise if it fails when trying to resize the buffer you lose the original pointer and leak the memory.

Additionally, in case of failure you need to release the existing buffer and prematurely return, otherwise you will hit a null deref in line 230

Reply:
Simplified the codeblock by not using a struct variable since the token string must always persist for a HTTP call.